docs/puzzles/day7.md

```diff
@@ -2,12 +2,327 @@
+```
+
+The crabmarine can move in two directions, forward and backwards.
+
+Let's translate our input into a list of `Crabmarine`:
+
+```scala
+val horizontalPositions = input.split(",").map(_.toInt).toList
+val crabmarines = horizontalPositions.map(horizontal => Crabmarine(horizontal))
+```
+
+A list of crabmarines can form a crabmada:
+
+```scala
+case class Crabmada(crabmarines: List[Crabmarine]):
+  // we don't want an empty list here!
+  require(crabmarines.nonEmpty)
+
+  def align() = ???
+```
+
+and it can have a method for aliging all its crabs.
+
+Now the tricky part, how can we figure out the horizontal position that
+minimizes the amount of fuel the crabs will use? We have only two possible
+moves, forwards and backwards, and we know for sure that all the edge crabs will
+need to move. Let's model out solution as doing one of two moves:
+
+- moving all the edge crabmarines with the current maximum horizontal position
+  backwards
+- moving all the edge crabmarines with the current minimal horizontal position
+  forwards
+
+In each interation we can move the crabmarines that use less fuel, which
+guarantees us that the result will be minimal at the end when all the crabs
+reach the same horizontal position.
+
+This can take a form of a recursive function:
+
+```scala
+  @tailrec
+  final def align(
+      situation: List[Crabmarine] = crabmarines,
+      fuelCost: Int = 0
+  ): Int =
+    val allTheSame = situation.forall(_.horizontal == situation.head.horizontal)
+    if allTheSame then fuelCost
+    else
+      val maxHorizontal = situation.maxBy(_.horizontal)
+      val minHorizontal = situation.minBy(_.horizontal)
+
+      val fuelCostForMax = situation.count {
+         crabmarine => crabmarine.horizontal == maxHorizontal.horizontal
+      }
+      val fuelCostForMin = situation.collect {
+        crabmarine => crabmarine.horizontal == minHorizontal.horizontal
+      }
+      if fuelCostForMax < fuelCostForMin then
+        val updated = situation.map { crabmarine =>
+          if crabmarine.horizontal == maxHorizontal.horizontal then
+            crabmarine.moveBackward()
+          else crabmarine
+        }
+        align(updated, fuelCost + fuelCostForMax)
+      else
+        val updated = situation.map { crabmarine =>
+          if crabmarine.horizontal == minHorizontal.horizontal then
+            crabmarine.moveForward()
+          else crabmarine
+        }
+        align(updated, fuelCost + fuelCostForMin)
+```
+
+First we check if all the crabmarines already align. If not we need to check
+what is the cost of moving all the crabs on each of the edges. We move all the
+ones on the edge that will take a up less fuel (less crabs to move). Then we
+invoke the function again with the updated positions of the crabmarines.
+
+Additionally, we use here a `@tailrec` annotation, which makes sure that out
+function can be translated by the compiler into an interative solution without
```

Comment:
```suggestion
function can be translated by the compiler into an imperative solution without
```

But maybe it’s more correct to say that it runs in constant stack space?

Reply:
I think you meant `iterative` rather than `imperative`

```suggestion
function can be translated by the compiler into an iterative solution without
```